_overviews/core/collections-generic-programming.md

```diff
@@ -0,0 +1,320 @@
+
+Several solutions can apply, depending on the desired level of genericity.
+
+### Consuming any *actual* collection
+
+Let’s start with the simplest case: consuming any collection that can be traversed.
+You don’t need to know the precise type of the collection,
+but just that it *is* a collection. This can be achieved by taking an `IterableOnce[A]`
+as parameter, or an `Iterable[A]` if you need more than one traversals. Here is an
+example that shows how to implement a `sumBy` operation that sums the elements of a
+collection after they have been transformed by a function:
+
+~~~ scala
+implicit class SumByOperation[A](coll: IterableOnce[A]) {
+  def sumBy[B](f: A => B)(implicit num: Numeric[B]): B = {
+    val it = coll.iterator
+    var result = f(it.next())
+    while (it.hasNext()) {
+      result = num.plus(result, it.next())
+    }
+    result
+  }
+}
+~~~
+
+We define the `sumBy` operation as an [implicit class](/overviews/core/implicit-classes.html) method so that
+it can be applied to a collection as if it was a method of that collection:
+
+~~~ scala
+case class User(name: String, age: Int)
+
+val users = Seq(User("Alice", 22), User("Bob", 20))
+
+println(users.sumBy(_.age)) // “42”
+~~~
+
+Unfortunately, this extension method does not work with values of type `String` and not
+even with `Array`. This is because these types are not part of the Scala collections
+hierarchy.
+
+### Consuming any type that is *like* a collection
+
+If we want the `sumBy` to work on any type that is *like* a collection, such as `String`
+and `Array`, we have to add another indirection level:
+
+~~~ scala
+import scala.collection.generic.IsIterableLike
+
+class SumByOperation[A](coll: IterableOnce[A]) {
+  def sumBy[B](f: A => B)(implicit num: Numeric[B]): B = ... // same as before
+}
+
+implicit def SumByOperation[Repr](coll: Repr)(implicit it: IsIterableLike[Repr]): SumByOperation[it.A] =
+  new SumByOperation[it.A](it(coll))
+~~~
+
+The type `IsIterableLike[Repr]` has implicit instances for all types `Repr` that can be converted
+to `IterableOps[A, Iterable, C]` (for some element type `A` and some collection type `C`). There are
+instances for actual collection types and also for `String` and `Array`.
+
+### Consuming a more specific collection than `Iterable`
+
+In some cases we want (or need) the receiver of the operation to be more specific than `Iterable`.
+For instance, some operations make sense only on `Seq` but not on `Set`.
+
+In such a case, again, the most straightforward solution would be to take as parameter a `Seq` instead
+of an `Iterable` or an `IterableOnce`, but this would work only with *actual* `Seq` values. If we want
+to support `String` and `Array` values we have to use `IsSeqLike` instead. `IsSeqLike` is similar to
+`IsIterableLike` but provides a conversion to `SeqOps[A, Iterable, C]` (for some types `A` and `C`).
+
+Note that in the case `Seq` as well as `Map`, using `IsSeqLike` and `IsMapLike`, respectively, is also
```

Comment:
The biggning is a bit hard to read here

Reply:
I rephrased this as follows:

> Using `IsSeqLike` is also required to make your operation work on `SeqView` values, because `SeqView`
does not extend `Seq`. Similarly, there is an `IsMapLike` type that makes operations work with
both `Map` and `MapView` values.